virttest/utils_qemu.py

```diff
@@ -19,10 +32,104 @@ def get_qemu_version(bin_path):
         raise OSError('Unable to get the version of qemu')
     return matches.groups()
+
+
+def get_machines_mapping(bin_path):
+    """
+    Return all machine type supported by qemu
+
+    :param bin_path: Path to qemu binary
+    :return: A dict of all machine types supported by qemu
+    """
+    output = _get_qemu_info(bin_path, r"-machine \?")
+    machines = re.findall(r"^([0-9a-zA-Z._-]+).*[^:]$", output, re.M)
+    machines_mapping = dict(zip(machines, machines))
+    has_alias_machines = re.findall(r"^([0-9a-zA-Z._-]+).*\(alias of (\S+)\)",
+                                    output, re.M)
+    machines_mapping.update(dict(has_alias_machines))
+    return machines_mapping
+
+
+def get_supported_machines(bin_path):
+    """
+    Return all machine type supported by qemu
+
+    :param bin_path: Path to qemu binary
+    :return: A list of all machine types supported by qemu
+    """
+    return list(get_machines_mapping(bin_path).keys())
+
+
+def get_correct_machine_type(bin_path, machine_type):
+    """
+    Return the correct machine type
+
+    :param bin_path: Path to qemu binary
+    :param machine_type: qemu machine type
+    :return: Correct machine type
+    """
+    try:
+        return get_machines_mapping(bin_path)[machine_type]
+    except KeyError:
+        raise KeyError("qemu does not support this machine type: %s"
+                       % machine_type)
+
+
+def get_supported_devices(bin_path, category=None):
+    """
+    Return all devices supported by qemu
+
+    :param bin_path: Path to qemu binary
+    :param category: device category (e.g. 'USB', 'Network', 'CPU')
+    :return: A set of devices supported by qemu
+    """
+    output = _get_qemu_info(bin_path, r"-device \?")
+    qemu_devices = {}
+    for device_info in output.split("\n\n"):
+        device_type = re.match(r"([A-Z]\S+) devices:", device_info).group(1)
+        qemu_devices[device_type] = set(re.findall(r'(?:name|alias) "(\S+)"',
+                                                   device_info))
+    if category:
+        return qemu_devices[category]
+    return set.union(*qemu_devices.values())
+
+
+def get_maximum_cpus(bin_path, machine_type):
+    """
+    Return max CPUs supported by specified machine type
+
+    :param bin_path: Path to qemu binary
+    :param machine_type: One machine type supported by qemu
+    :raise ValueError: If unable to get that
+    :return: Maximum value of vCPU
+    """
+    invalid_maximum = 2 ** 31 - 1
+    correct_machine_type = get_correct_machine_type(bin_path, machine_type)
+    output = _get_qemu_info(bin_path, r"-machine %s -smp maxcpus=%d" %
+                            (correct_machine_type, invalid_maximum))
+    searches = re.search(r"'%s' is (\d+)" % correct_machine_type, output)
+    if searches is None:
+        raise ValueError("Could not get the max CPUs supported by machine '%s'"
+                         % correct_machine_type)
+    return int(searches.group(1))
+
+
+def get_cpu_models(bin_path):
```

Comment:
Seems this function implements same thing with function `extract_qemu_cpu_models`.

Reply:
you are right, we need to move it from virttest.cpu into utils_qemu